src/modules/uORB/uORBDeviceNode.cpp

```diff
@@ -299,13 +289,7 @@ uORB::DeviceNode::ioctl(cdev::file_t *filp, int cmd, unsigned long arg)
 		}
 
 	case ORBIOCUPDATED:
-#ifndef __PX4_NUTTX
-		lock();
```

Comment:
Why would locking still be necessary?
https://github.com/PX4/Firmware/blob/88f7d2d05654c75ed771277e8947adb9a87a3537/src/modules/uORB/uORBDeviceNode.cpp#L495-L497

Subscriptions aren't shared between threads, and the only time a poll_notify (different thread) would be in here is if the subscription owning thread was in a poll setup (not doing other work).

Reply:
To ensure the access to `_generation` is atomic and ordered as we want to.
Long-term we will need to introduce atomic variables, then we can handle this better.

Reply:
The generation is a word and I'm still not seeing any potential ordering issues. Can you think of an example?

Reply:
There's no guarantee that an access to a word-sized variable is atomic. While it is true for ARM, I don't want to depend on platform-specifics. Especially not at a place like uORB.

A probably surprising reordering effect can be that in an orb publish, _generation is increased before the memcpy happens, and an orb_check will then already see an update (w/o a CS), but the data is not updated yet. In that case an orb_copy would still wait because it's protected by another critical section, but it should make clear that the argumentation is not straight-forward.
C++ only guarantees consistency within a single thread, consistency between threads you have to take care of explicitly (meaning the compiler and the CPU can arbitrarily reorder and cache things as long as the result is the same for an individual thread).

And don't get me wrong, I want to reduce the amount of locking as well.

Reply:
@bkueng - 

Please correct me if I have the wrong understanding of how this works: As I see it the sd is held by a single reader task. The only shared and accessed variable is _generation. It may be updated from an ISR or task. _generation is defined volatile. So there is no caching in registers other then the instruction sequence load,load,compare or load inc store and those are interruptible but also not critical. There is only one writer. The effect load inc store can be mitigated with a delta check on `_generation` comparing the values before and after memcpy. If interrupted recopy the data.  The load, load compare will have the consequence of the ISR firing and interrupting the load, load as  `load ISR load`  all this would do is delays the delivery.  This happens regularly. 

I do not think that bus locking or d-caching is an issue here. If it is we should use the https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html to access the shared data and not use interrupt disabling.

I step in ASM mode a lot the I have seen amazing optimization performed, by the compiler. Have you ever seen the compiler reorder a function calls sequence and a volatile?

Reply:
@davids5 Your correct (except there can be multiple writers), but you are assuming a specific instruction set or architecture here (i.e. ARM), and relying on assumptions that are not guaranteed by the C/C++ language. If you do that you end up with non-portable implementation.

Regarding `volatile`: it's really not the right thing to use for synchronization, it's been designed for different purposes. It might often work, but I really want to get away from it for synchronization. GCC's atomic builtins are exactly the right thing to use for that. I've been meaning to add those for some time but never got around to do it yet. You find a good article in the linux kernel docs about that: https://www.kernel.org/doc/html/latest/process/volatile-considered-harmful.html.

> Have you ever seen the compiler reorder a function calls sequence and a volatile?

No, but do you really want to rely on something just because you've never seen it? Also compilers get better and better.

Reply:
I disagree given the current state of uORB the change was made against (both read()/write() in critical sections, and volatile _generation), however I don't think it's worth arguing over now as I was planning to come back sometime after https://github.com/PX4/Firmware/pull/11176 goes in to see how much of the Subscriber side can be made lock free.

Please review the updated PR.